### ISSUE TITLE
Error 401

### ISSUE THREAD

**Issue by pluc** [soundcloud/api#110]
I am aware that the previous issue here is also called 'HTTP Error 401' but given that it feels like a different issue I have elected to create a separate task.

We got the notice about the upcoming changes to the API so we had to move to a [different library](https://github.com/danae/soundcloud-php). Having done so, and using a library that implements the upcoming changes, our requests no longer work.

Now I've contacted support and the first time I got a:

> We have made a quick change on our backend for the key on your account

which I have no idea how to interpret. Once that didn't change the output, the subsequent support person replied:

> It looks like your API credentials got revoked, but we've reinstated them since. 

which again I have no idea how that could happen, but here we are.

The error we get from the payload is not helpful (because you are not providing a reason), and support told us to post here.

## Error 401 on Authenticated requests

```
Upload failed because of: The request "POST tracks?client_id=hhhhh&oauth_token=hhhhhhh" returned with HTTP status code 401: {"code":401,"message":"","link":"https://developers.soundcloud.com/docs/api/explorer/open-api","status":"401 - Unauthorized","errors":[],"error":null}
```

That request failed with a 401 when passing this in the headers:
```
array (
'Authorization' => 'OAuth oauth token here',
)
```

### Endpoint(s):
* `POST tracks`

### Scope(s):
* No idea what this means

### Steps to reproduce:
```bash
curl -X POST -H 'Authorization: OAuth oauth_token' -F @file_i_assume 'https://apiurl/tracks?client_id=<your clientid>&oauth_token=oauth token'
```

### Expected behaviour:
All we want is that the POST request works as per the documentation - right now it is giving us a 401 when it should let us in. The result should be for the track to be uploaded.

### Actual behaviour:
401 Unauthorized

**Comment by dasha-kobzeva** [comment]
Hi @pluc,

per our documentation for this specific endpoint https://developers.soundcloud.com/docs/api/explorer/open-api#/tracks/post_tracks the request is required to provide track info in a specific format with particular fields being mandatory.

Also, I can see that you send the auth header, oauth_token, and client_id at the same time. This is incorrect and might be causing the issue. Only Auth header is accepted. Please remove client_id and oauth_token from your request url.

**Comment by pluc** [comment]
Well I am getting this issue again by using a snippet that is [generated by your documentation](https://developers.soundcloud.com/docs/api/explorer/open-api#/tracks/post_tracks):

```bash
$ curl -X POST "https://api.soundcloud.com/tracks" -H  "accept: application/json; charset=utf-8" -H  "Authorization: OAuth mytoken" -H  "Content-Type: multipart/form-data" -F "track[downloadable]=true" -F "track[sharing]=private" -F "track[streamable]=true" -F "track[asset_data]=@mytrack.mp3;type=audio/mpeg" -F "track[commentable]=true" -F "track[title]=testTrack"
{"code":401,"message":"","link":"https://developers.soundcloud.com/docs/api/explorer/open-api","status":"401 - Unauthorized","errors":[],"error":null}
```

**Comment by rahulrumalla** [comment]
We're here to help. I understand you may be frustrated, but it would be more useful to us if instead you could provide more information to diagnose. 

Have you checked if your token expired? The access tokens are short lived.

**Comment by pluc** [comment]
The token is obtained a few seconds before the attempt at uploading a track.

**Comment by mgoodfellow** [comment]
Are you using a user connect flow login? i.e. via the connect screen and swap code for oauth token? You can't upload a track via the client_credentials flow (i.e. client_id / client_secret swapped for an oauth token) as it doesn't have a user context

**Comment by pluc** [comment]
@mgoodfellow No, this is for a scheduled CLI app that uploads things in the background so there is no user interactivity and uses `client_credentials` grant_type. If that's true, is the only way to upload a track to involve a user prompt now?

**Comment by mgoodfellow** [comment]
Well, sort of - you have to have the user go through that login at least once first (via the connect screen flow), then once you have the refresh token, you could generate an access_token as required.

In this case, your "app" doesn't have a user context, so it can't upload to a user. There is a link about this somewhere, but I cannot find it quickly. The client_credentials flow access_token gives you access to the non-user based public API routes.

**Comment by mgoodfellow** [comment]
Found it - it was linked here: https://github.com/soundcloud/api/issues/107#issuecomment-899478892

**Comment by pluc** [comment]
It's a cli application, this used to work without user interaction and we can't easily implement it, so that's a little problematic.

**Comment by pluc** [comment]
@mgoodfellow are you saying that we can no longer automatically upload to soundcloud without user interaction? Our use case is a scheduled job that runs in the background and backs up local media and uploads it to soundcloud. We can't really insert periodic user interaction in that, do you have a suggestion?

**Comment by mgoodfellow** [comment]
At some point, yes, you must authenticate as a user in order to upload as that user. But once that user interaction has taken place you can upload from a backend process if required. I guess in the old way you provided an account email/password but support for that has since been dropped. I can't solve your specific use case, but I would say that once you have generated the refresh token you can continue to use that as required on your backend, and generate a new access_token/refresh_token pair to consume as required in the future, therefore you wouldn't require "periodic" user interaction

**Comment by dasha-kobzeva** [comment]
Dear @pluc,
We are sorry for the difficulty here. 

As @mgoodfellow has mentioned the password grant type was recently deprecated. It introduces a number of vulnerabilities as you are exposing your user credentials.

Authorization Code flow is the method that allows you to interact with our API on a user's behalf. Yes, it requires manual action from a user in the browser, but that can be done only once, and as suggested `refresh_token` flow can be used afterward.

We have also made a note regarding our error codes, thank you.
